core/blockchain.go

```diff
@@ -1639,15 +1626,24 @@ func (bc *BlockChain) insertChain(chain types.Blocks, verifySeals bool) (int, er
 		// Update the metrics touched during block validation
 		accountHashTimer.Update(statedb.AccountHashes) // Account hashes are complete, we can mark them
 		storageHashTimer.Update(statedb.StorageHashes) // Storage hashes are complete, we can mark them
-
 		blockValidationTimer.Update(time.Since(substart) - (statedb.AccountHashes + statedb.StorageHashes - triehash))
 
 		// Write the block to the chain and get the status.
 		substart = time.Now()
-		status, err := bc.writeBlockWithState(block, receipts, logs, statedb, false)
-		atomic.StoreUint32(&followupInterrupt, 1)
-		if err != nil {
-			return it.index, err
+		var status WriteStatus
+		if !setHead {
+			// Don't set the head, only insert the block
+			err := bc.writeBlockWithState(block, receipts, logs, statedb)
+			atomic.StoreUint32(&followupInterrupt, 1)
+			if err != nil {
+				return it.index, err
+			}
+		} else {
+			status, err = bc.writeBlockAndSetHead(block, receipts, logs, statedb, false)
+			atomic.StoreUint32(&followupInterrupt, 1)
+			if err != nil {
+				return it.index, err
+			}
 		}
```

Comment:
This looks a bit quirky, since in one clause, you define a new `err`, and in the other, you use the existing one. Is there some particular reason for that? Otherwise, you can just do

```suggestion
		if !setHead {
			// Don't set the head, only insert the block
			err = bc.writeBlockWithState(block, receipts, logs, statedb)
		} else {
			status, err = bc.writeBlockAndSetHead(block, receipts, logs, statedb, false)
		}
		atomic.StoreUint32(&followupInterrupt, 1)
		if err != nil {
			return it.index, err
		}

```